src/bootstrap/config.rs

```diff
@@ -564,6 +577,13 @@ impl Config {
         set(&mut config.configure_args, build.configure_args);
         set(&mut config.local_rebuild, build.local_rebuild);
         set(&mut config.print_step_timings, build.print_step_timings);
+        config.stage = config.stage.or(build.stage);
+        config.doc_stage = build.doc_stage.or(config.stage);
+        config.test_stage = build.test_stage.or(config.stage);
+        config.build_stage = build.build_stage.or(config.stage);
+        config.bench_stage = build.bench_stage.or(config.stage);
+        config.install_stage = build.install_stage.or(config.stage);
+        config.dist_stage = build.dist_stage.or(config.stage);
```

Comment:
I think this is probably wrong. Why do we need to know the stage for multiple subcommands? I think we should be able to have just config.stage, with a similar match on subcommand as you modified above. config.stage would then be a `u32`, non-optional.

Reply:
If there's only `config.stage`, not `config.build_stage` or anything like that, then the only choices are 'make all the default stages the same' or 'use the provided defaults'. That leaves out case 3 above.

Am I misunderstanding your suggestion?

Reply:
The toml still contains per-subcommand values, but we should normalize those as early as possible in bootstrap if possible. Definitely we should not expose *both* stage and foo_stage, because we'll use the wrong one.

Reply:
I'm fine with removing `stage = ...` in config.toml, but I'm confused how that allows `config.stage` to be normalized early or remove these `.or()`s. The place that uses the `Option` is `builder.rs`:
```rust
builder.rs
574-        // CI should always run stage 2 builds, unless it specifically states otherwise
575-        #[cfg(not(test))]
576:        if build.config.stage.is_none() && build.ci_env != crate::CiEnv::None {
```
Should I remove that assert? There are definitely some cases when CI tests stage 0, like `doc --stage 0`.

Reply:
I would prefer that we remove `config.*_stage`. I mean that if we have `config.stage` and `config.doc_stage` we need documentation as to which one should be used when, and I'm pretty confident we'll get it wrong. If we just have one "category" of stage then there's no question which one should be used.

Reply:
Got it, took me a while 😆 how does it look now?